websocket/polygon.go

```diff
@@ -0,0 +1,228 @@
+package polygonws
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/polygon-io/client-go/websocket/models"
+	"go.uber.org/zap"
+)
+
+// todo: add reconnect logic
+
+type Client struct {
+	apiKey string
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	conn   *websocket.Conn
+	rQueue chan []byte
+	wQueue chan []byte
+
+	log Logger
+}
+
+// todo: might want to separate Connect logic out from New
+func New(config Config) (*Client, error) {
+	if config.APIKey == "" {
+		return nil, errors.New("API key is required")
+	}
+
+	url := fmt.Sprintf("wss://%v.polygon.io/%v", string(config.Feed), string(config.Market))
+
+	if config.Log == nil {
+		config.Log = zap.NewNop().Sugar()
+	}
+
+	// todo: is this default dialer sufficient? might want to let user pass in a context so they can cancel the dial
+	conn, res, err := websocket.DefaultDialer.Dial(url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial server: %w", err)
+	} else if res.StatusCode != 101 {
+		return nil, errors.New("server failed to switch protocols")
+	}
+
+	conn.SetReadLimit(maxMessageSize)
+	if err := conn.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+		return nil, fmt.Errorf("failed to set read deadline: %w", err)
+	}
+	conn.SetPongHandler(func(string) error {
+		return conn.SetReadDeadline(time.Now().Add(pongWait))
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c := &Client{
+		apiKey: config.APIKey,
+		ctx:    ctx,
+		cancel: cancel,
+		conn:   conn,
+		rQueue: make(chan []byte, 10000),
+		wQueue: make(chan []byte, 100),
+		log:    config.Log,
+	}
+
+	go c.read()
+	go c.write()
+	go c.process()
+
+	return c, nil
+}
+
+// todo: Subscribe, Unsubscribe, etc
+
+func (c *Client) Close() error {
+	c.cancel()
+	err := c.conn.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""), time.Now().Add(writeWait))
```

Comment:
There's technically a race condition between this and the writer process. Although practically I'm not sure if it will really matter, is this something that you'd considered and decided to go with?

Reply:
it's a bit unclear to me but i _think_ calling WriteControl concurrently with WriteMessage is allowed (see this [issue](https://github.com/gorilla/websocket/issues/83#issuecomment-147906076)). old issue but i've seen this kind of pattern elsewhere too. going to dig some more and regardless, might refactor this to use the wQueue instead

Reply:
leaving a todo comment here so i'll remember to verify later, going to merge as is for now